File: sparsegs/scores.py
"""Gene-set scoring methods, implemented against the published definitions.

Two of the four methods are rank-based and can be evaluated from a
:class:`~sparsegs.rankcache.RankCache`; the other two are expression-based and
need the matrix.  All four are here so that a calibration result can be reported
as a property of *the analysis*, not of one particular scoring function.

Methods
-------
aucell
    Aibar et al., *Nat. Methods* 2017.  Area under the recovery curve over the
    top ``max_rank`` ranks, normalised by ``n_genes_in_set * max_rank``.
ucell
    Andreatta & Carmona, *Comput. Struct. Biotechnol. J.* 2021.  Mann-Whitney U
    of the set's capped ranks, normalised to ``[0, 1]``.
score_genes
    The scanpy default (Satija lab).  Mean expression of the set minus the mean
    of expression-bin-matched control genes.  Produces no null distribution.
module_score
    Seurat ``AddModuleScore``.  As above, with controls averaged per expression
    bin rather than pooled.
ssgsea
    Barbie et al., *Nature* 2009.  Weighted ECDF difference over the full
    ranking.  Included for completeness; needs the matrix, not the cache.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
import scipy.sparse as sp

__all__ = ["aucell", "ucell", "score_genes", "module_score", "ssgsea",
           "score_all"]


# ----------------------------------------------------------------------
# rank-based methods
# ----------------------------------------------------------------------
def aucell(cache, genes, max_rank=None):
    """AUCell scores from a rank cache.

    Parameters
    ----------
    cache : RankCache
    genes : sequence of str
        Gene set.  Genes absent from the cache are dropped.
    max_rank : int, optional
        Rank ceiling.  Defaults to ``ceil(0.05 * cache.n_genes_total)``, the
        AUCell default ranking fraction.

    Returns
    -------
    numpy.ndarray
        One score per cell, in ``[0, 1]``.
    """
    present = cache.present(genes)
    if len(present) < 1:
        raise ValueError("none of the requested genes are in the cache")
    if max_rank is None:
        max_rank = int(np.ceil(0.05 * cache.n_genes_total))
    max_rank = int(max_rank)

    r = cache.ranks(present)
    contrib = np.where((r > 0) & (r <= max_rank), max_rank - r + 1, 0)
    return contrib.sum(axis=1) / (len(present) * max_rank)


def ucell(cache, genes, r_max=1500, normalisation="v2"):
    """UCell scores from a rank cache.

    Parameters
    ----------
    cache : RankCache
        The cache ceiling must be at least ``r_max``.
    genes : sequence of str
    r_max : int
        Rank cap.  The UCell documentation recommends setting this to the median
        number of detected genes per cell; the package default is 1500, which is
        appropriate for 10x data with typical depth but not for shallow data.
    normalisation : {'v2', 'v1'}
        ``'v2'`` uses ``U_max = k * r_max - k (k + 1) / 2``, the corrected
        maximum introduced in UCell v2.  ``'v1'`` uses the original paper's
        ``U_max = k * r_max``.

    Returns
    -------
    numpy.ndarray
        One score per cell, in ``[0, 1]`` up to a small negative floor.

    Notes
    -----
    The published cap is ``r' = r_max + 1`` for ranks beyond ``r_max``, so the
    largest value the U statistic can take is ``k (r_max + 1) - k (k + 1) / 2``,
    while the v2 normaliser is ``k * r_max - k (k + 1) / 2``.  The two differ by
    ``k``, so the score has a floor at ``-1 / (r_max - (k + 1) / 2)`` rather than
    exactly zero.  At the package default (``r_max = 1500``, ``k = 8``) the floor
    is ``-7e-4`` and invisible; it grows as ``r_max`` is lowered towards the
    median detected-gene count, and reaches ``-0.018`` at ``r_max = 60`` with
    ``k = 10``.  Scores are returned unclipped so that they reproduce the
    reference implementation exactly.
    """
    present = cache.present(genes)
    if len(present) < 1:
        raise ValueError("none of the requested genes are in the cache")
    r_max = int(r_max)
    if r_max > cache.ceiling:
        raise ValueError(
            f"cache ceiling {cache.ceiling} is below r_max {r_max}; rebuild the "
            f"cache with a ceiling of at least {r_max}")

    k = len(present)
    r = cache.ranks(present)
    # Ranks above the cap -- including the BEYOND sentinel -- become r_max + 1,
    # which is exactly the published cap.
    capped = np.where((r > 0) & (r <= r_max), r, r_max + 1)
    u_stat = capped.sum(axis=1) - k * (k + 1) / 2.0
    u_max = k * r_max if normalisation == "v1" else k * r_max - k * (k + 1) / 2.0
    return 1.0 - u_stat / u_max


# ----------------------------------------------------------------------
# expression-based methods
# ----------------------------------------------------------------------
def _expression_bins(mean_expression, n_bins):
    """Assign genes to equal-count bins of mean expression.

    Returns an integer bin label per gene, plus one boolean mask per bin.
    """
    ranks = pd.Series(mean_expression).rank(method="first")
    labels = pd.qcut(ranks, n_bins, labels=False).to_numpy()
    labels = np.where(np.isfinite(labels), labels, 0).astype(int)
    n_used = labels.max() + 1
    masks = [labels == b for b in range(n_used)]
    return labels, masks


def _mean_expression(X, positions):
    """Column means of a sparse matrix restricted to ``positions``."""
    sub = X[:, positions]
    return np.asarray(sub.mean(axis=1)).ravel()


def score_genes(X, genes, gene_set, ctrl_size=50, n_bins=25, seed=0):
    """scanpy's ``score_genes``: set mean minus expression-matched control mean.

    Parameters
    ----------
    X : scipy.sparse matrix
        Cells x genes expression matrix.
    genes : sequence of str
    gene_set : sequence of str
    ctrl_size : int
        Number of control genes drawn per gene of the set.
    n_bins : int
        Number of expression bins.
    seed : int

    Returns
    -------
    numpy.ndarray
        One score per cell.  Not bounded; higher means the set is more expressed
        than its expression-matched controls.
    """
    X = sp.csr_matrix(X)
    genes = np.asarray(genes)
    pos = {g: i for i, g in enumerate(genes)}
    target = np.array([pos[g] for g in gene_set if g in pos], dtype=np.int64)
    if target.size == 0:
        raise ValueError("none of the genes in gene_set are present")

    mean_expression = np.asarray(X.mean(axis=0)).ravel()
    labels, masks = _expression_bins(mean_expression, n_bins)
    in_set = np.zeros(len(genes), dtype=bool)
    in_set[target] = True

    rng = np.random.default_rng(seed)
    controls = []
    for t in target:
        candidates = np.flatnonzero(masks[labels[t]] & ~in_set)
        if candidates.size == 0:
            candidates = np.flatnonzero(~in_set)
        take = min(ctrl_size, candidates.size)
        controls.append(rng.choice(candidates, size=take, replace=False))
    control = np.concatenate(controls)

    return _mean_expression(X, target) - _mean_expression(X, control)


def module_score(X, genes, gene_set, ctrl_size=100, n_bins=24, seed=0):
    """Seurat's ``AddModuleScore``: per-bin control means, subtracted.

    The difference from :func:`score_genes` is that controls are averaged within
    each expression bin and each target gene has its own bin's control mean
    subtracted, rather than pooling every control gene into one average.
    """
    X = sp.csr_matrix(X)
    genes = np.asarray(genes)
    pos = {g: i for i, g in enumerate(genes)}
    target = np.array([pos[g] for g in gene_set if g in pos], dtype=np.int64)
    if target.size == 0:
        raise ValueError("none of the genes in gene_set are present")

    mean_expression = np.asarray(X.mean(axis=0)).ravel()
    labels, masks = _expression_bins(mean_expression, n_bins)
    in_set = np.zeros(len(genes), dtype=bool)
    in_set[target] = True

    rng = np.random.default_rng(seed)
    bin_control_mean = {}
    for b, mask in enumerate(masks):
        if not mask.any():
            continue
        candidates = np.flatnonzero(mask & ~in_set)
        if candidates.size == 0:
            candidates = np.flatnonzero(~in_set)
        take = min(ctrl_size, candidates.size)
        chosen = rng.choice(candidates, size=take, replace=False)
        bin_control_mean[b] = _mean_expression(X, chosen)

    score = np.zeros(X.shape[0], dtype=float)
    for t in target:
        score += (_mean_expression(X, np.array([t]))
                  - bin_control_mean.get(int(labels[t]), 0.0))
    return score / target.size


def ssgsea(X, genes, gene_set):
    """Weighted ECDF difference (Barbie et al. 2009) for one gene set.

    Implemented for completeness and used only where matrices are small enough
    that a full ranking pass is affordable.
    """
    X = sp.csr_matrix(X)
    genes = np.asarray(genes)
    pos = {g: i for i, g in enumerate(genes)}
    target = np.array([pos[g] for g in gene_set if g in pos], dtype=np.int64)
    if target.size == 0:
        raise ValueError("none of the genes in gene_set are present")

    dense = np.asarray(X.todense(), dtype=np.float32)
    n_cells, n_genes = dense.shape
    order = np.argsort(-dense, axis=1, kind="stable")
    rank_of = np.empty((n_cells, n_genes), dtype=np.int32)
    np.put_along_axis(rank_of, order, np.arange(1, n_genes + 1, dtype=np.int32)[None, :],
                      axis=1)

    in_set = np.zeros(n_genes, dtype=bool)
    in_set[target] = True
    out = np.empty(n_cells, dtype=float)
    for i in range(n_cells):
        ranks = np.sort(rank_of[i, target])
        values = np.abs(dense[i, target])
        weight = values / max(values.sum(), 1e-12)
        hit = np.cumsum(weight)
        miss_total = n_genes - target.size
        miss = np.zeros(target.size, dtype=float)
        # Number of non-set genes ranked above each set member, normalised.
        for j, r in enumerate(ranks):
            miss[j] = (r - 1 - j) / max(miss_total, 1)
        out[i] = hit.sum() / target.size - miss.sum() / target.size
    return out


# ----------------------------------------------------------------------
# convenience
# ----------------------------------------------------------------------
def score_all(cache, X, genes, gene_set, ucell_r_max=1500, seed=0):
    """Evaluate every method on one gene set, returning a tidy frame.

    Parameters
    ----------
    cache : RankCache
    X : scipy.sparse matrix
        Needed for the expression-based methods; pass ``None`` to skip them.
    genes : sequence of str
    gene_set : sequence of str
    ucell_r_max : int
    seed : int

    Returns
    -------
    pandas.DataFrame
        Columns ``method``, ``cell``, ``score``.
    """
    frames = []
    for name, values in (
            ("AUCell", aucell(cache, gene_set)),
            ("UCell", ucell(cache, gene_set, r_max=ucell_r_max))):
        frames.append(pd.DataFrame({"method": name,
                                    "cell": np.arange(len(values)),
                                    "score": values}))
    if X is not None:
        for name, fn in (("score_genes", score_genes),
                         ("module_score", module_score)):
            values = fn(X, genes, gene_set, seed=seed)
            frames.append(pd.DataFrame({"method": name,
                                        "cell": np.arange(len(values)),
                                        "score": values}))
    return pd.concat(frames, ignore_index=True)

File: r-package/sparseGenSetCal/R/scores.R
#' Gene-set scores, implemented against their published definitions
#'
#' Two of the methods are rank-based and evaluate from a `RankCache`; the other
#' two are expression-based and need the matrix.  All four are here so that a
#' calibration result can be reported as a property of *the analysis* rather
#' than of one particular scoring function.
#'
#' @references
#' Aibar S. et al. \emph{Nature Methods} 14, 1083-1086 (2017).
#'
#' Andreatta M., Carmona S.J. \emph{Computational and Structural Biotechnology
#' Journal} 19, 3796-3798 (2021).
#'
#' @name scores
#' @keywords internal
NULL

#' AUCell scores
#'
#' Area under the recovery curve over the top `max_rank` ranks, normalised by
#' `length(genes) * max_rank`.
#'
#' @param cache A `RankCache`.
#' @param genes Character vector.  Genes absent from the cache are dropped.
#' @param max_rank Rank ceiling; defaults to `ceil(0.05 * cache$n_genes_total)`.
#'
#' @return Numeric vector of scores in \eqn{[0, 1]}, one per cell.
#'
#' @examples
#' X <- make_synthetic_matrix(n_cells = 100, n_genes = 500, seed = 2)
#' cache <- rank_cache(X$matrix, X$genes, ceiling = 40)
#' head(aucell(cache, X$genes[1:10]))
#'
#' @export
aucell <- function(cache, genes, max_rank = NULL) {
    pos <- cache$positions(genes)
    if (length(pos) == 0) {
        stop("none of the requested genes are in the cache", call. = FALSE)
    }
    if (is.null(max_rank)) {
        max_rank <- as.integer(ceiling(0.05 * cache$n_genes_total))
    }
    max_rank <- as.integer(max_rank)
    if (max_rank > cache$ceiling) {
        # A rank above the ceiling was never stored, so the score silently
        # omits it: the answer is a lower bound rather than an AUCell score.
        # This is easy to do by accident, because the default ceiling is
        # ceil(0.05 * G) and the cache may have been built with less.
        warning(sprintf(
            "max_rank %d is above the cache ceiling %d, whose ranks were not stored; the score omits them and is a lower bound. Rebuild the cache with ceiling >= %d.",
            max_rank, cache$ceiling, max_rank), call. = FALSE)
    }
    r <- cache$clipped[, pos, drop = FALSE]
    contrib <- ifelse(r > 0 & r <= max_rank, max_rank - r + 1, 0)
    rowSums(contrib) / (length(pos) * max_rank)
}

#' UCell scores
#'
#' Mann-Whitney U of the set's capped ranks, normalised to \eqn{[0, 1]}.  The
#' UCell documentation recommends setting `r_max` to the median number of
#' detected genes per cell, which [depth_report()] reports; the package default
#' of 1500 suits 10x data with typical depth and is far too high for shallow
#' data.
#'
#' @inheritParams aucell
#' @param r_max Rank cap.
#' @param normalisation `"v2"` uses \eqn{U_{max} = k r_{max} - k(k+1)/2}, the
#'   corrected maximum introduced in UCell v2; `"v1"` uses the original paper's
#'   \eqn{U_{max} = k r_{max}}.
#'
#' @return Numeric vector of scores.
#'
#' @details
#' The published cap is \eqn{r' = r_{max} + 1} for ranks beyond the cap, so the
#' largest value the U statistic can take is
#' \eqn{k(r_{max}+1) - k(k+1)/2}, while the v2 normaliser is
#' \eqn{k r_{max} - k(k+1)/2}.  The two differ by \eqn{k}, so the score floors
#' at \eqn{-1/(r_{max} - (k+1)/2)} rather than exactly zero: invisible at
#' \eqn{r_{max} = 1500}, but \eqn{-0.018} at \eqn{r_{max} = 60} with \eqn{k = 10}.
#' Scores are returned unclipped so that they reproduce the reference
#' implementation exactly.
#'
#' @export
ucell <- function(cache, genes, r_max = 1500, normalisation = "v2") {
    pos <- cache$positions(genes)
    if (length(pos) == 0) {
        stop("none of the requested genes are in the cache", call. = FALSE)
    }
    r_max <- as.integer(r_max)
    if (r_max > cache$ceiling) {
        stop(sprintf(
            "cache ceiling %d is below r_max %d; rebuild the cache with a ceiling of at least %d",
            cache$ceiling, r_max, r_max), call. = FALSE)
    }
    k <- length(pos)
    r <- cache$clipped[, pos, drop = FALSE]
    capped <- ifelse(r > 0 & r <= r_max, r, r_max + 1)
    u_stat <- rowSums(capped) - k * (k + 1) / 2
    u_max <- if (identical(normalisation, "v1")) {
        k * r_max
    } else {
        k * r_max - k * (k + 1) / 2
    }
    1 - u_stat / u_max
}

#' Mean expression of a gene set, minus expression-matched controls
#'
#' The scanpy default.  For each gene of the set, `ctrl_size` control genes are
#' drawn from the same expression bin, and the score is the mean expression of
#' the set minus the mean of the pooled controls.  It produces no null
#' distribution on its own; it is here so that a calibration result can be
#' checked across scoring rules.
#'
#' @param X Cells x genes matrix.
#' @param genes Gene names, one per column of `X`.
#' @param gene_set Character vector.
#' @param ctrl_size Number of control genes drawn per gene of the set.
#' @param n_bins Number of expression bins.
#' @param seed Integer.
#'
#' @return Numeric vector of scores, one per cell.  Not bounded; higher means
#'   the set is more expressed than its expression-matched controls.
#'
#' @export
score_genes <- function(X, genes, gene_set, ctrl_size = 50, n_bins = 25,
                        seed = 0) {
    genes <- as.character(genes)
    pos <- match(intersect(as.character(gene_set), genes), genes)
    if (length(pos) == 0) {
        stop("none of the genes in gene_set are present", call. = FALSE)
    }
    X <- as_dgCMatrix(X)
    mean_expression <- as.numeric(Matrix::colMeans(X))
    labels <- expression_bins(mean_expression, n_bins)

    set.seed(seed)
    controls <- integer(0)
    for (t in pos) {
        candidates <- which(labels == labels[t] & !(seq_along(genes) %in% pos))
        if (length(candidates) == 0) {
            candidates <- which(!(seq_along(genes) %in% pos))
        }
        controls <- c(controls,
                      sample(candidates, min(ctrl_size, length(candidates))))
    }
    as.numeric(Matrix::rowMeans(X[, pos, drop = FALSE]) -
               Matrix::rowMeans(X[, controls, drop = FALSE]))
}

#' @rdname scores
#' @keywords internal
expression_bins <- function(values, n_bins) {
    cut(rank(values, ties.method = "first"),
        breaks = unique(stats::quantile(
            seq_along(values) / (length(values) + 1),
            probs = seq(0, 1, length.out = n_bins + 1))),
        include.lowest = TRUE, labels = FALSE)
}

#' Score the same set by every method the framework supports
#'
#' @param cache A `RankCache`.
#' @param X The expression matrix the cache was built from, or `NULL` to skip
#'   the expression-based methods.
#' @param gene_set Character vector.
#' @param genes Gene names for the columns of `X`.
#' @param ucell_r_max Rank cap for UCell.
#' @param seed Integer, for the expression-matched controls.
#'
#' @return A data frame with columns `method`, `cell` and `score`.
#'
#' @export
score_all <- function(cache, X, genes, gene_set, ucell_r_max = 1500,
                      seed = 0) {
    frames <- list(
        data.frame(method = "AUCell", cell = seq_len(nrow(cache$clipped)),
                   score = aucell(cache, gene_set)),
        data.frame(method = "UCell", cell = seq_len(nrow(cache$clipped)),
                   score = ucell(cache, gene_set, r_max = ucell_r_max)))
    if (!is.null(X)) {
        frames[[length(frames) + 1L]] <- data.frame(
            method = "score_genes", cell = seq_len(nrow(cache$clipped)),
            score = score_genes(X, genes, gene_set, seed = seed))
    }
    do.call(rbind, frames)
}
